src/lamp_py/performance_manager/alerts.py

```diff
@@ -0,0 +1,478 @@
+                ("active_period_start_timestamp", pyarrow.uint64()),
+                (
+                    "active_period_end",
+                    pyarrow.timestamp("ms", tz="America/New_York"),
+                ),
+                ("active_period_end_timestamp", pyarrow.uint64()),
+                ("informed_entity.route_id", pyarrow.string()),
+                ("informed_entity.route_type", pyarrow.int8()),
+                ("informed_entity.direction_id", pyarrow.int8()),
+                ("informed_entity.stop_id", pyarrow.string()),
+                ("informed_entity.facility_id", pyarrow.string()),
+                ("informed_entity.activities", pyarrow.string()),
+            ]
+        )
+
+    def existing_id_timestamp_pairs(self) -> pandas.DataFrame:
+        """
+        get all unique alert id / last modified timestamp pairs in the local
+        parquet file, a key that can be used to identify alerts that have
+        already been processed.
+        """
+        columns = ["id", "last_modified_timestamp"]
+        if self.local_exists:
+            existing_alerts = pq.read_table(
+                self.local_path, columns=columns
+            ).to_pandas()
+            existing_alerts = existing_alerts.drop_duplicates()
+            return existing_alerts
+
+        return pandas.DataFrame(columns=columns)
+
+    def append_new_records(self, alerts: pandas.DataFrame) -> None:
+        """
+        append alerts to the end of the local parquet file using batches to keep memory usage lower
+        """
+        self.new_data = True
+        alerts = alerts.reset_index(drop=True)
+        alerts_table = pyarrow.Table.from_pandas(
+            alerts, schema=self.parquet_schema
+        )
+
+        # if there is no local file, write the alerts to the local file and exit
+        if not self.local_exists:
+            with pq.ParquetWriter(
+                self.local_path, schema=self.parquet_schema
+            ) as writer:
+                writer.write_table(alerts_table)
+
+            self.local_exists = True
+            return
+
+        # write new alerts to a new file and then join them with the local file in batches
+        new_path = os.path.join("/tmp", "new_alerts.parquet")
+        joined_path = os.path.join("/tmp", "joined_alerts.parquet")
+
+        with pq.ParquetWriter(new_path, schema=self.parquet_schema) as writer:
+            writer.write_table(alerts_table)
+
+        joined_batches = pd.dataset(
+            [
+                pd.dataset(self.local_path),
+                pd.dataset(new_path),
+            ]
+        ).to_batches()
+
+        with pq.ParquetWriter(
+            joined_path, schema=self.parquet_schema
+        ) as writer:
+            for batch in joined_batches:
+                writer.write_batch(batch)
+
+        # the joined file becomes the local file and we can delete the new and
+        # joined files
+        os.replace(joined_path, self.local_path)
+        os.remove(new_path)
```

Comment:
i didn't add all of the logging on this call yet as i don't think it'll be anywhere near as intensive as the ALL_RT file.